### PR TITLE
Feature: Support imports when using Protobuf files

### PR DESCRIPTION
**Pull request by joschi**
- [ ] **Breaking change?** (if so, please describe the impact and migration path for existing application instances)

**What changes did you make?** (Give an overview)

The changes in https://github.com/provectus/kafka-ui/pull/2874 added initial support for using more than 1 Protobuf file in Kafka UI in absence of a proper schema registry.

This change is building upon that functionality to support more complex scenarios in which there are multiple Protobuf files being used and not all of them are explicitly listed (for example imports).

It's using the already present Wire library from Square to do the heavy lifting and create a comprehensive schema from all Protobuf files and directories listed in the Kafka UI configuration.

See also https://github.com/provectus/kafka-ui/pull/2874#pullrequestreview-1168052798.

**Is there anything you'd like reviewers to focus on?**

**How Has This Been Tested?** (put an "x" (case-sensitive!) next to an item)

- [ ] No need to
- [ ] Manually (please, describe, if necessary)
- [x] Unit checks
- [ ] Integration checks
- [x] Covered by existing automation

**Checklist** (put an "x" (case-sensitive!) next to all the items, otherwise the build will fail)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. **ENVIRONMENT VARIABLES**)
- [x] My changes generate no new warnings (e.g. Sonar is happy)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged

Check out [Contributing](https://github.com/provectus/kafka-ui/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/provectus/kafka-ui/blob/master/CODE-OF-CONDUCT.md)

**A picture of a cute animal (not mandatory but encouraged)**

Piglet

### CODE REVIEW

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serdes/builtin/ProtobufFileSerde.java [hunk 1]
```diff
@@ -70,9 +78,33 @@ public void configure(PropertyResolver serdeProperties,
...
+    schemaLoader.setLoadExhaustively(true);
+    schemaLoader.setPermitPackageCycles(true);
+    schemaLoader.initRoots(locations, List.of());
+
+    final Schema schema;
+    try {
+      schema = schemaLoader.loadSchema();
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
```
Comment (iliax) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/serdes/builtin/ProtobufFileSerde.java:97]: please extract to static package-access `ProtobufFileSerde::loadProtobufSchema` method and use it in tests also. Maybe put `toLocation` logic there too

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serdes/builtin/ProtobufFileSerde.java [hunk 2]
```diff
@@ -127,6 +159,17 @@ void configure(
...
+  @Nullable
+  private static Location toLocation(Path path) {
+    if (Files.isDirectory(path)) {
+      return Location.get(path.toString());
+    }
+    if (Files.isRegularFile(path)) {
+      return Location.get(path.getParent().toString(), path.getFileName().toString());
+    }
+    return null;
+  }
```
Comment (iliax) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/serdes/builtin/ProtobufFileSerde.java:171]: just a question: so now we can set directory with `.proto` files? If so, will it be scanned recursively?

**Review by iliax — CHANGES_REQUESTED** [review]

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serdes/builtin/ProtobufFileSerde.java (hunk 1 above)
Comment (joschi) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/serdes/builtin/ProtobufFileSerde.java]: Done.

https://github.com/provectus/kafka-ui/pull/3039/commits/b7a91fbd9f37930f56aa9d1fde1bf33af5675ec0

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serdes/builtin/ProtobufFileSerde.java (hunk 2 above)
Comment (joschi) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/serdes/builtin/ProtobufFileSerde.java:174]: @iliax Yes, that's correct. Directories will be scanned recursively.

The Square Wire library even supports reading Protobuf schemas from ZIP (and JAR) archives.

I have updated the documentation and added explicit support for reading Protobuf files from archives, too.

➡️ https://github.com/provectus/kafka-ui/pull/3039/commits/9b226e3fe9aebe701307bf341db1f07440ef9a9e

**Comment by Haarolean** [comment]
@joschi  hey, thanks for one more contribution. We're currently in a code freeze stage, we'll be able to properly review and merge your PR right after we do a release.

**Comment by joschi** [comment]
@Haarolean Thanks for the update!

It would've been nice to get this into Kafka UI 0.5.0, but I also understand the need for properly testing a non-moving target before a release.

Let me know if I can help you moving this forward. 👍

**Comment by joschi** [comment]
@iliax I've rebased the PR to the latest `master` branch. Is there anything more for me to do _here_ before it can be merged?

> In next release we'll probably change app building policy (will use -cp rather than executable jar), so this won't be a problem.

How can I help you do this? Can you please refer me to the respective ticket(s) if nobody is working on it yet? 😄

**Comment by Haarolean** [comment]
@joschi we'll get to this soon, sorry :)

**Comment by iliax** [comment]
@joschi we deciede to rearrage code from your PR to
- remove new dependency, since it depends on kotlin runtilme, need to be synched with other deps and adds 8mb to imadge size
- refactor existing code in serde 

here is PR https://github.com/provectus/kafka-ui/pull/3357
if you have time, we would be glad if you take a look on it

**Comment by iliax** [comment]
closing due to comment above

**Pull request closed without merging**